### PR TITLE
update to handle beta removals in 1.22

### PR DESCRIPTION
**Pull request by deads2k**
Some APIs are removed in 1.22.  This PR handles that removal in advance of a beta being tagged at which point these APIs are automatically no longer served.

/kind cleanup
/priority important-soon
/assign @liggitt 

```release-note
NONE
```

### CODE REVIEW

**Comment by wojtek-t** [comment]
/test pull-kubernetes-e2e-gce-100-performance 

To check if we will gather logs.

**Comment by BenTheElder** [comment]
/test pull-kubernetes-e2e-kind

flake? that's an incredibly [slow run that timed out](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-kind/1375133285402808320). will look closer later.

**Comment by wojtek-t** [comment]
@deads2k - https://github.com/kubernetes/test-infra/pull/21568 should address the first problem - we will see what happens once that merges :)

**Comment by wojtek-t** [comment]
OK - that PR merged, let's see what happens:

/test pull-kubernetes-e2e-gce-100-performance

**Comment by wojtek-t** [comment]
@deads2k - currently I'm getting:
```
error: no matches for kind "Ingress" in version "networking.k8s.io/v1beta1"
```
in addon-manager

When looking into open-api, it seems that indeed networking.k8s.io/v1beta1/Ingress is no longer served.
But I don't really understand why - Ingress doesn't seem to have 'delete' annotation:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/networking/v1beta1/types.go#L25
Why it's not served?

**Comment by liggitt** [comment]
Beta types are automatically no longer served three releases after being deprecated. Ingress  is only served under v1 in 1.22

**Comment by deads2k** [comment]
> Beta types are automatically no longer served three releases after being deprecated unless explicitly extended by a longer delete annotation version. Ingress is only served under v1 in 1.22

In addition, 1.18 goes EOL at the end of April.  Could we just update the test to only run on 1.19 and higher, deciding that it is better to run on current releases than finish out the last month of 1.18?

**Comment by BenTheElder** [comment]
Haven't had a chance to look deeper yet and probably won't for a while (going OOO soon), but https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-kind/1375311672427679744 has a pretty substantial runtime increase seemingly due to a lot of tests timing out.

Are CSI tests using a removed API?

**Comment by liggitt** [comment]
> Are CSI tests using a removed API?

https://github.com/kubernetes/kubernetes/issues/99957?

**Comment by wojtek-t** [comment]
After upgrades to scalability test, they are now failing with:
```
[measurement call TestMetrics - TestMetrics error: [action gather failed for SystemPodMetrics measurement: restart counts violation: RestartCount(l7-lb-controller-e2e-99840-95a39-master, l7-lb-controller)=3, want <= 0]]
```

I guess this probably expected, as we didn't yet upgrade the ingress-controller to the new version that doesn't use v1beta1 CRD.

@deads2k - FYI

**Comment by deads2k** [comment]
/retest

?

**Comment by liggitt** [comment]
pull-kubernetes-e2e-gce-ubuntu-containerd clusters will start now 🎉 

#99957 looks like the next biggest blocker for passing e2e runs.

**Comment by wojtek-t** [comment]
The l7lb controller is still crashlooping in performance tests - it is the only remaining problem with performance tests.

**Comment by jprzychodzen** [comment]
`pull-kubernetes-e2e-gce-100-performance` is blocked on kubernetes/ingress-gce#1414

**Comment by BenTheElder** [comment]
I expected at least `pull-kubernetes-e2e-kind` to improve after https://github.com/kubernetes/kubernetes/issues/99957

**Comment by msau42** [comment]
Looking at one of the failures, it's not failing due to CSINode anymore. The volume is successfully created and mounted on the node. It appears like creating the pod sandbox is timing out:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-kind/1385282039325921280/artifacts/logs/kind-worker2/kubelet.log
```
Apr 22 17:46:55 kind-worker2 kubelet[246]: E0422 17:46:55.813989     246 kuberuntime_sandbox.go:68] "Failed to create sandbox for pod" err="rpc error: code = DeadlineExceeded desc = context deadline exceeded" pod="ephemeral-5793/inline-volume-tester-xxtsx"
Apr 22 17:46:55 kind-worker2 kubelet[246]: E0422 17:46:55.814018     246 kuberuntime_manager.go:790] "CreatePodSandbox for pod failed" err="rpc error: code = DeadlineExceeded desc = context deadline exceeded" pod="ephemeral-5793/inline-volume-tester-xxtsx"
```

This continues until the test times out at 17:57

**Comment by liggitt** [comment]
artifacts like https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-kind/1385282039325921280/artifacts/_sig-storage_CSI_mock_volume/CSIServiceAccountToken/token_should_be_plumbed_down_when_csiServiceAccountTokenEnabled_true/csi-mockplugin-attacher-0-csi-attacher.log show csi-mockplugin-attacher-0/csi-attacher using v1beta1 resources still:

```
E0422 17:42:09.117495       1 reflector.go:156] k8s.io/client-go/informers/factory.go:135: Failed to list *v1beta1.VolumeAttachment: the server could not find the requested resource
E0422 17:42:09.117531       1 reflector.go:156] k8s.io/client-go/informers/factory.go:135: Failed to list *v1beta1.CSINode: the server could not find the requested resource
I0422 17:42:10.117730       1 reflector.go:188] Listing and watching *v1beta1.VolumeAttachment from k8s.io/client-go/informers/factory.go:135
I0422 17:42:10.118844       1 reflector.go:188] Listing and watching *v1beta1.CSINode from k8s.io/client-go/informers/factory.go:135
E0422 17:42:10.121076       1 reflector.go:156] k8s.io/client-go/informers/factory.go:135: Failed to list *v1beta1.CSINode: the server could not find the requested resource
E0422 17:42:10.121232       1 reflector.go:156] k8s.io/client-go/informers/factory.go:135: Failed to list *v1beta1.VolumeAttachment: the server could not find the requested resource
I0422 17:42:11.121331       1 reflector.go:188] Listing and watching *v1beta1.CSINode from k8s.io/client-go/informers/factory.go:135
I0422 17:42:11.122380       1 reflector.go:188] Listing and watching *v1beta1.VolumeAttachment from k8s.io/client-go/informers/factory.go:135
E0422 17:42:11.123323       1 reflector.go:156] k8s.io/client-go/informers/factory.go:135: Failed to list *v1beta1.CSINode: the server could not find the requested resource
E0422 17:42:11.123916       1 reflector.go:156] k8s.io/client-go/informers/factory.go:135: Failed to list *v1beta1.VolumeAttachment: the server could not find the requested resource
```

**Comment by msau42** [comment]
> artifacts like https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-kind/1385282039325921280/artifacts/_sig-storage_CSI_mock_volume/CSIServiceAccountToken/token_should_be_plumbed_down_when_csiServiceAccountTokenEnabled_true/csi-mockplugin-attacher-0-csi-attacher.log show csi-mockplugin-attacher-0/csi-attacher using v1beta1 resources still:

Doh csi mock driver sidecars also needs to be updated. Working on a PR...

**Comment by msau42** [comment]
Mock driver updated, ready for a retest...

**Comment by liggitt** [comment]
weirdly, on 4/21, the cluster would [start and pass all but 46 tests](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-gce-ubuntu-containerd/1384864155151896576). The v1.11.0 glbc image [crashlooped](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-gce-ubuntu-containerd/1384864155151896576/artifacts/e2e-7a2398ea27-a7d53-master/glbc.log) for a while with this error:

```
I0421 14:17:45.606685       1 crd.go:76] ServiceNetworkEndpointGroup CRD is Established.
F0421 14:17:45.730816       1 main.go:174] app.NewNamer(ctx.KubeClient, "", "") = the server could not find the requested resource (get ingresses.extensions)
```

but then appeared to run in a crippled state where the informer could not sync:
```
I0421 14:22:59.855401       1 crd.go:76] ServiceNetworkEndpointGroup CRD is Established.
I0421 14:22:59.942422       1 namer.go:129] Using uid = "c54346f792cfb64b" saved in ConfigMap
...
```

in the latest runs, the cluster will [not even start](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-gce-ubuntu-containerd/1387401050482282496), and the v1.11.0 glbc image is [crashlooping](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/99840/pull-kubernetes-e2e-gce-ubuntu-containerd/1387401050482282496/artifacts/e2e-6983be5d93-a7d53-master/glbc.log) with the same error:

```
I0428 14:16:31.782166       1 crd.go:76] ServiceNetworkEndpointGroup CRD is Established.
F0428 14:16:31.793940       1 main.go:174] app.NewNamer(ctx.KubeClient, "", "") = the server could not find the requested resource (get ingresses.extensions)
```

**Comment by wojtek-t** [comment]
@deads2k - scalability test passed

File: pkg/controlplane/storageversionhashdata/data.go [hunk 1]
```diff
@@ -63,53 +63,37 @@ var GVRToStorageVersionHash = map[string]string{
 	"batch/v1/jobs":          "mudhfqk/qZY=",
```
Comment (liggitt) [inline pkg/controlplane/storageversionhashdata/data.go:63]: interestingly, this file doesn't have CRD or APIService hashes

File: staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go [hunk 2]
```diff
@@ -118,6 +118,9 @@ func (e *resourceExpirationEvaluator) shouldServe(gv schema.GroupVersion, versio
 }
 
 func (e *resourceExpirationEvaluator) ShouldServeForVersion(majorRemoved, minorRemoved int) bool {
+	if minorRemoved == 22 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go:121]: change to `if e.currentMinor == 22 && minorRemoved == 22` if you want to mimic the logic that will apply once 1.22 reaches beta

File: test/cmd/legacy-script.sh [hunk 3]
```diff
@@ -850,7 +850,7 @@ runTests() {
     kube::test::get_object_assert 'clusterroles -l test-cmd=auth' "{{range.items}}{{$id_field}}:{{end}}" 'testing-CR:'
 
     failure_message=$(! kubectl auth reconcile "${kube_flags[@]}" -f test/fixtures/pkg/kubectl/cmd/auth/rbac-v1beta1.yaml 2>&1 )
-    kube::test::if_has_string "${failure_message}" 'only rbac.authorization.k8s.io/v1 is supported'
+    kube::test::if_has_string "${failure_message}" 'no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"'
```
Comment (liggitt) [inline test/cmd/legacy-script.sh:853]: `reconcile` doesn't need restmapper info, it pins to v1 RBAC types and makes v1 API calls explicitly.

I think this might be a better change:

```diff
diff --git a/staging/src/k8s.io/kubectl/pkg/cmd/auth/reconcile.go b/staging/src/k8s.io/kubectl/pkg/cmd/auth/reconcile.go
index 3ead45bf644..5151310ace3 100644
--- a/staging/src/k8s.io/kubectl/pkg/cmd/auth/reconcile.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/auth/reconcile.go
@@ -138,6 +138,7 @@ func (o *ReconcileOptions) Complete(cmd *cobra.Command, f cmdutil.Factory, args
 		NamespaceParam(namespace).DefaultNamespace().
 		FilenameParam(enforceNamespace, o.FilenameOptions).
 		Flatten().
+		Local().
 		Do()
 
 	if err := r.Err(); err != nil {
```

File: staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go [hunk 4]
```diff
@@ -118,6 +118,9 @@ func (e *resourceExpirationEvaluator) shouldServe(gv schema.GroupVersion, versio
 }
 
 func (e *resourceExpirationEvaluator) ShouldServeForVersion(majorRemoved, minorRemoved int) bool {
+	if minorRemoved <= 22 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go:121]:
```suggestion
	if e.currentMinor == 22 && minorRemoved <= 22 {
```

File: test/integration/etcd/data.go [hunk 5]
```diff
@@ -168,14 +168,6 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
 		},
 		// --
```
Comment (liggitt) [inline test/integration/etcd/data.go:170]: same here, revert this change, it's not required for integration tests, which aren't build-version aware

Comment (deads2k) [reply test/integration/etcd/data.go:170]: > same here, revert this change, it's not required for integration tests, which aren't build-version aware

I think integration tests should become version aware (somehow) to ensure that we're testing what we actually want to enforce.  How we make that happen, I'm open to suggestions for.

Comment (liggitt) [reply test/integration/etcd/data.go:170]: For this test specifically, which checks the persisted path and version of every served type, it should exercise every API the server is possibly capable of serving. I'd expect it to exercise these until the storage for these versions is actually removed.

File: staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go [hunk 6]
```diff
@@ -118,6 +118,10 @@ func (e *resourceExpirationEvaluator) shouldServe(gv schema.GroupVersion, versio
 }
 
 func (e *resourceExpirationEvaluator) ShouldServeForVersion(majorRemoved, minorRemoved int) bool {
+	// TODO remove sometime after 1.22 hits beta.  This allows us to latch all progress and stop serving while we're still alpha
+	if e.currentMinor == 22 && minorRemoved == 22 {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go:125]: now that we're plumbing good version info here, we can get the same result by making alpha handling strict below, right?

```golang
// TODO remove sometime after 1.22 hits beta. This allows us to latch all progress and stop serving while we're still alpha
// if e.isAlpha && e.strictRemovedHandlingInAlpha { // don't serve in alpha if we want strict handling
return false
// }
```

that seems preferable to teaching k8s.io/apiserver about `1.22` specifically

File: test/integration/framework/master_utils.go [hunk 7]
```diff
@@ -271,7 +271,8 @@ func NewIntegrationTestMasterConfigWithOptions(opts *MasterConfigOptions) *contr
 
 // MasterConfigOptions are the configurable options for a new integration test master config.
 type MasterConfigOptions struct {
-	EtcdOptions *options.EtcdOptions
+	EtcdOptions                    *options.EtcdOptions
+	ServeRemovedAPIsOneMoreRelease bool
```
Comment (liggitt) [inline test/integration/framework/master_utils.go:275]: was this used?

Comment (deads2k) [reply test/integration/framework/master_utils.go]: > was this used?

In one iteration it was attempted, but the wiring for the etcd storage test is actually via options, not via config.  The kube-apiserver wiring has never been fully untangled.  Though after removing the ssh tunnels it will get a little easier.

**Comment by liggitt** [comment]
weird unit test failure with no details I could see:

```
=== FAIL: vendor/k8s.io/apiextensions-apiserver/pkg/registry/customresource  (0.00s)
FAIL	k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/registry/customresource	10.197s
```

**Comment by liggitt** [comment]
/approve

lgtm once the unit test issue is resolved

**Comment by liggitt** [comment]
#101953 failure

/retest

**Comment by liggitt** [comment]
needs rebase
